walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,120 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.HashMap;
+import java.util.ArrayList;
+import java.util.Collections;
 
+/** 
+* Returns a Collection of TimeRange objects that span any available meeting times for all people in the MeetingRequest request.
+* For example, if 4pm - 8pm is free for a meeting, 
+* the Time Range object will represent that entire chunk of time rather than 4pm - 5pm, even if the duration of meeting is 1 hour. 
+*/
 public final class FindMeetingQuery {
+  public static final int MIN_INCREMENT = 10;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    long mtngDuration = request.getDuration();
+    Collection<String> attendees = request.getAttendees();
+    Collection<Event> allEvents = events;
+
+    HashMap<Integer, Integer> mandatoryEvents = makeMandatoryEventMap(allEvents, attendees);
+    Collection<TimeRange> meetingTimes = findMeetingTimes(mtngDuration, mandatoryEvents);
+    return meetingTimes;
+  }
+
+
+  /** 
+  * Create a hash map <event start time, event end time> that ONLY contains meetings where 1 or more people from attendees must attend. 
+  **/
+  public HashMap<Integer, Integer> makeMandatoryEventMap (Collection<Event> allEvents, Collection<String> attendees) {
+    HashMap<Integer, Integer> mandatoryEvents = new HashMap<Integer, Integer>();
+
+    for (Event event : allEvents) {
+      //people who must attend 
+      Set<String> people = event.getAttendees();
+      //check if any people in meeting request is present in this set
+      for (String person : attendees) {
```

Comment:
I don't think you actually need to iterate through every single attendee, right?  You just need to determine if this event contains any of the provided attendees, and if so, add the event time to your map.

You may be able to remove some of this nesting by using [java streams anyMatch](https://www.geeksforgeeks.org/stream-anymatch-java-examples/) to find out of this event is potentially conflicting with the request.  anyMatch is nice because it's short-circuiting and will terminate if it finds a match, not having to go through all attendees.

boolean eventHasAttendee = attendees.stream().anyMatch(attendee -> event.getAttendees().contains(attendee));

if (eventHasAttendee) {
  addMandatoryEvent(event);
}

And then your addMandatory event function can add the event to your map (unless it's shorter than the existing entry).

(As a side note, code with lots of nesting has the potential to introduce bugs since it makes it hard to follow the conditional branching that may or may not occur)